lib/Dialect/ESI/runtime/cpp/include/esi/Types.h

```diff
@@ -153,6 +387,69 @@ class StructType : public Type {
     return size;
   }
 
+  std::pair<bool, std::string> isValid(const std::any &obj) const override {
+    try {
+      auto structData = std::any_cast<std::map<std::string, std::any>>(obj);
+
+      if (structData.size() != fields.size()) {
+        return {false, "struct has " + std::to_string(structData.size()) +
+                           " fields, expected " +
+                           std::to_string(fields.size())};
+      }
+
+      for (const auto &[fieldName, fieldType] : fields) {
+        auto it = structData.find(fieldName);
+        if (it == structData.end()) {
+          return {false, "missing field '" + fieldName + "'"};
+        }
+
+        auto [fieldValid, fieldReason] = fieldType->isValid(it->second);
+        if (!fieldValid) {
+          return {false, "invalid field '" + fieldName + "': " + fieldReason};
+        }
+      }
+      return {true, ""};
+    } catch (const std::bad_any_cast &) {
+      return {false, "must be std::map<std::string, std::any>"};
+    }
+  }
+
+  MessageData serialize(const std::any &obj) const override {
+    auto [valid, reason] = isValid(obj);
+    if (!valid) {
+      throw std::runtime_error("Invalid object for struct type: " + reason);
+    }
+
+    auto structData = std::any_cast<std::map<std::string, std::any>>(obj);
+    std::vector<uint8_t> result;
```

Comment:
For performance, might it be a good idea to reserve space for the entire struct?